lib/transfer-memory-allocator.js

```diff
@@ -0,0 +1,548 @@
+const PRIVATE = Symbol("private");
+/**
+ * Keeps track of memory usable for transfers, and allocates available 
+ * memory as needed/when requested by transfers
+ */
+class TransferMemoryBuffer {
+    /**
+     * Constructs a transfer memory buffer
+     * @param {*} suggestedSize Suggested size for the memory reserved 
+     * for buffering transfer data, in bytes. The memory allocator will
+     * try to allocate memory with size `suggestedSize`, but if not 
+     * enough memory is available, 80% of available memory will be 
+     * allocated instead
+     */
+    constructor(suggestedSize) {
+        this[PRIVATE] = {};
+
+        let bufferSize;
+        if (suggestedSize) {
+            console.log(`Suggested Transfer Memory Buffer pool size is ${suggestedSize} bytes`);
+            bufferSize = parseInt(suggestedSize, 10);
+        } else {
+            console.log(`Using default size for Transfer Memory Buffer: ${DEFAULT_TRANSFER_BUFFER_POOL_SIZE} bytes`);
+            bufferSize = DEFAULT_TRANSFER_BUFFER_POOL_SIZE;
+        }
+
+        if (!bufferSize || !Number.isInteger(bufferSize)) {
+            throw new Error("Suggested Transfer Memory Buffer total suggested size must be an integer");
+        }
+        if (bufferSize <= 0) {
+            throw new Error("Transfer Memory Buffer total suggested size must be larger than 0");
+        }
+        
+
+        const availableFreeMemory = os.freemem();
+        const highMemoryMark = Math.floor(availableFreeMemory * MAX_PERCENTAGE_ALLOWED_MEMORY_USE);
+        if (bufferSize > highMemoryMark) {
+            // use 80% of available memory
+            this[PRIVATE].totalSize = Math.min(highMemoryMark, MAX_BUFFER_SIZE);
+            console.log(`Not enough memory available (${availableFreeMemory} bytes available, ${bufferSize} bytes requested) to use suggested size for Transfer Memory Buffer pool size. Using ${MAX_PERCENTAGE_ALLOWED_MEMORY_USE * 100}% of available memory instead: ${this[PRIVATE].totalSize} bytes as Transfer Memory Buffer pool size`);
+        } else {
+            console.log(`Using suggested size of ${bufferSize} bytes for Transfer Memory Buffer pool size`);
+            this[PRIVATE].totalSize = Math.min(bufferSize, MAX_BUFFER_SIZE);
+        }
+        this[PRIVATE].availablePooledBufferSize = this[PRIVATE].totalSize;
+
+        // create buffer that will be used as buffer pool
+        this[PRIVATE].pooledBuffer = Buffer.allocUnsafe(this[PRIVATE].totalSize);
+        this[PRIVATE].minIndex = BUFFER_POOL_ALLLOC_START_INDEX;
+        this[PRIVATE].maxIndex = BUFFER_POOL_ALLLOC_START_INDEX + this[PRIVATE].totalSize - 1;
+
+        // create list to keep track of allocated buffer blocks from the larger buffer `pooledBuffer`
+        this[PRIVATE].allocatedBlocks = yallist.create([]);
+
+        // allocations waiting for enough memory to be available
+        this[PRIVATE].pendingAllocations = yallist.create([]);
+    }
```

Comment:
`It has support for pending allocations, which could potentially cause a problem in the transfer algorithm.`
Could you elaborate which issues you see? Because a bunch of tests cover possible issues.